plugins/image-prioritizer/class-image-prioritizer-video-tag-visitor.php

```diff
@@ -123,4 +128,48 @@ private function preload_poster_image( OD_Tag_Visitor_Context $context ): bool {
 
 		return true;
 	}
+
+	/**
+	 * Adjust `autoplay` and `preload` values for videos outside initial viewport.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_Tag_Visitor_Context $context Tag visitor context, with the cursor currently at an embed block.
+	 * @return bool Whether the tag should be tracked in URL metrics.
+	 */
+	private function lazy_load_videos( OD_Tag_Visitor_Context $context ): bool {
+		$processor = $context->processor;
+
+		$xpath = $processor->get_xpath();
+
+		$intersection_ratio = $context->url_metric_group_collection->get_element_max_intersection_ratio( $xpath );
+
+		if ( $intersection_ratio > 0 ) {
```

Comment:
I wonder if `$intersection_ratio > 0` is the right condition here because it could very well be that the URL Metrics are not fully populated. For example, it could be that if URL Metrics were gathered for mobile visitors in which the video is not in the viewport. For desktop visitors, the video could be in the viewport but no URL Metrics have been collected yet for it. Here the `$intersection_ratio` would be `0.0` since mobile the video is not visible on any mobile visitors but this would result in the video being lazy-loaded which would end up hurting desktop visitors for whom the video is visible in the viewport.

I realize this may be a problem for lazy-loaded images and embeds as well.

To deal with this, the lazy-loading optimization should perhaps depend on the smallest viewport group (mobile) being populated as well as the largest group (desktop). But if a site only ever gets mobile or desktop visitors, then this could indefinitely delay the optimization from being applied, which isn't good. Since this is a wholistic problem/concern/question for videos, images, and embeds, it probably isn't something needing to be addressed in this PR and we can open an issue to consider further.

Related: https://github.com/WordPress/performance/pull/1595#discussion_r1803421782

For prioritizing the loading of LCP images the situation is a bit different since we aren't dependent on the one `fetchpriority=high` attribute being on the `IMG` tag. Instead, we're able to add preload links with `fetchpriority=high` that also have media queries to ensure that the populated URL Metrics will ensure the image gets loaded properly.

Also related https://github.com/WordPress/performance/pull/1404